version: "3.8"

services:
  web:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: fastapi_app
    depends_on:
      db:
        condition: service_healthy
      redis:
        condition: service_started
    ports:
      - "8000:8000"
    env_file:
      - .env
    networks:
      - my_network

  worker:
    build:
      context: .
      dockerfile: Dockerfile.worker
    container_name: celery_worker
    env_file:
      - .env
    depends_on:
      db:
        condition: service_healthy
      redis:
        condition: service_started
    networks:
      - my_network

  redis:
    image: redis:7-alpine
    container_name: redis_broker
    ports:
      - "6379:6379"
    networks:
      - my_network

  db:
    image: postgres:15-alpine
    container_name: postgres_db
    environment:
      POSTGRES_DB: "${POSTGRES_DB}"
      POSTGRES_USER: "${POSTGRES_USER}"
      POSTGRES_PASSWORD: "${POSTGRES_PASSWORD}"
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U ${POSTGRES_USER} -d ${POSTGRES_DB} || exit 1"]
      interval: 5s
      timeout: 5s
      retries: 5
    ports:
      - "5432:5432"
    networks:
      - my_network


networks:
  my_network:
    driver: bridge
